File: edge/internal/contract/descriptor.go
package contract

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"math"
	"strconv"
	"strings"
	"unicode"
)

const MaxDescriptorBytes = 1024 * 1024

type DescriptorSnapshot struct {
	SchemaVersion      uint32             `json:"schema_version"`
	EdgeNodeID         string             `json:"edge_node_id"`
	LedgerEpoch        string             `json:"ledger_epoch"`
	DescriptorRevision uint64             `json:"descriptor_revision"`
	Complete           bool               `json:"complete"`
	Devices            []DescriptorDevice `json:"devices"`
	Signals            []DescriptorSignal `json:"signals"`
}

type DescriptorDevice struct {
	SystemID   string  `json:"system_id"`
	Identifier *string `json:"identifier,omitempty"`
	State      string  `json:"state"`
	ModelID    *string `json:"model_id,omitempty"`
}

type DescriptorSignal struct {
	SeriesKey      string  `json:"series_key"`
	SystemID       string  `json:"system_id"`
	MeasurementKey string  `json:"measurement_key"`
	ChannelIndex   *int32  `json:"channel_index"`
	Variant        string  `json:"variant"`
	Unit           *string `json:"unit"`
	ValueType      string  `json:"value_type"`
}

type SeriesIdentity struct {
	SystemID       string
	MeasurementKey string
	ChannelIndex   *int32
	Variant        string
}

func ParseSeriesKey(seriesKey string) (SeriesIdentity, error) {
	var noIdentity SeriesIdentity
	if strings.IndexFunc(seriesKey, unicode.IsControl) >= 0 {
		return noIdentity, fmt.Errorf("invalid series_key: control character")
	}
	parts := strings.Split(seriesKey, ":")
	if len(parts) != 4 || !validUUID(parts[0]) || !validMeasurementKey(parts[1]) || parts[3] == "" {
		return noIdentity, fmt.Errorf("invalid series_key: expected canonical four-part identity")
	}
	var channelIndex *int32
	if parts[2] != "na" {
		parsed, err := strconv.ParseInt(parts[2], 10, 32)
		if err != nil || parsed < 0 || strconv.FormatInt(parsed, 10) != parts[2] {
			return noIdentity, fmt.Errorf("invalid series_key: non-canonical channel")
		}
		value := int32(parsed)
		channelIndex = &value
	}
	return SeriesIdentity{
		SystemID:       parts[0],
		MeasurementKey: parts[1],
		ChannelIndex:   channelIndex,
		Variant:        parts[3],
	}, nil
}

func validMeasurementKey(key string) bool {
	if len(key) == 0 || len(key) > 64 {
		return false
	}
	segmentStart := true
	for index := 0; index < len(key); index++ {
		character := key[index]
		if character == '.' {
			if segmentStart {
				return false
			}
			segmentStart = true
			continue
		}
		if segmentStart {
			if character < 'a' || character > 'z' {
				return false
			}
			segmentStart = false
			continue
		}
		if (character < 'a' || character > 'z') &&
			(character < '0' || character > '9') && character != '_' {
			return false
		}
	}
	return !segmentStart
}

func DecodeDescriptorSnapshot(payload []byte) (DescriptorSnapshot, error) {
	var snapshot DescriptorSnapshot
	if len(payload) > MaxDescriptorBytes {
		return snapshot, descriptorInvalid("snapshot exceeds encoded byte limit")
	}
	if err := decodeOneStrict(payload, &snapshot); err != nil {
		return snapshot, fmt.Errorf("decode descriptor snapshot: %w", err)
	}
	if err := snapshot.Validate(); err != nil {
		return snapshot, err
	}
	return snapshot, nil
}

func (snapshot DescriptorSnapshot) Validate() error {
	if snapshot.SchemaVersion != 2 || !snapshot.Complete {
		return descriptorInvalid("only complete descriptor schema version 2 is supported")
	}
	if err := validateTopicSegment("edge_node_id", snapshot.EdgeNodeID); err != nil {
		return err
	}
	if err := validateIdentityComponent("ledger_epoch", snapshot.LedgerEpoch); err != nil {
		return err
	}
	if snapshot.DescriptorRevision < 1 || snapshot.DescriptorRevision > math.MaxInt64 {
		return descriptorInvalid("descriptor_revision must be positive")
	}
	if snapshot.Devices == nil || snapshot.Signals == nil {
		return descriptorInvalid("devices and signals arrays are required")
	}

	deviceIDs := make(map[string]struct{}, len(snapshot.Devices))
	for _, device := range snapshot.Devices {
		if !validUUID(device.SystemID) {
			return descriptorInvalid("device system_id is not a UUID")
		}
		if _, duplicate := deviceIDs[device.SystemID]; duplicate {
			return descriptorInvalid("duplicate device system_id")
		}
		deviceIDs[device.SystemID] = struct{}{}
		if device.State != "quarantined" && device.State != "active" && device.State != "retired" {
			return descriptorInvalid("unsupported device state")
		}
		if device.Identifier != nil && !validDisplayText(*device.Identifier, 64, false) {
			return descriptorInvalid("invalid device identifier")
		}
		if device.ModelID != nil && !validModelID(*device.ModelID) {
			return descriptorInvalid("invalid device model_id")
		}
	}

	seriesKeys := make(map[string]struct{}, len(snapshot.Signals))
	for _, signal := range snapshot.Signals {
		if !validUUID(signal.SystemID) {
			return descriptorInvalid("signal system_id is not a UUID")
		}
		if _, exists := deviceIDs[signal.SystemID]; !exists {
			return descriptorInvalid("signal references an unknown device")
		}
		identity, err := ParseSeriesKey(signal.SeriesKey)
		if err != nil || identity.SystemID != signal.SystemID ||
			identity.MeasurementKey != signal.MeasurementKey ||
			identity.Variant != signal.Variant ||
			!sameChannelIndex(identity.ChannelIndex, signal.ChannelIndex) {
			return descriptorInvalid("series_key does not match signal identity")
		}
		if _, duplicate := seriesKeys[signal.SeriesKey]; duplicate {
			return descriptorInvalid("duplicate signal series_key")
		}
		seriesKeys[signal.SeriesKey] = struct{}{}
		if signal.ValueType != "float" && signal.ValueType != "int" && signal.ValueType != "bool" && signal.ValueType != "record" {
			return descriptorInvalid("unsupported signal value_type")
		}
		if signal.Unit != nil && !validDisplayText(*signal.Unit, 128, true) {
			return descriptorInvalid("invalid signal unit")
		}
	}
	return nil
}

func validModelID(value string) bool {
	if len(value) == 0 || len(value) > 64 || value[0] < 'a' || value[0] > 'z' {
		return false
	}
	afterSeparator := false
	for index := 1; index < len(value); index++ {
		character := value[index]
		if (character >= 'a' && character <= 'z') ||
			(character >= '0' && character <= '9') {
			afterSeparator = false
			continue
		}
		if (character == '-' || character == '_' || character == '.') &&
			!afterSeparator {
			afterSeparator = true
			continue
		}
		return false
	}
	return !afterSeparator
}

func sameChannelIndex(left, right *int32) bool {
	if left == nil || right == nil {
		return left == nil && right == nil
	}
	return *left == *right
}

func (snapshot DescriptorSnapshot) ContentSHA256() ([sha256.Size]byte, error) {
	if err := snapshot.Validate(); err != nil {
		return [sha256.Size]byte{}, err
	}
	payload, err := json.Marshal(snapshot)
	if err != nil {
		return [sha256.Size]byte{}, err
	}
	return sha256.Sum256(payload), nil
}

func validUUID(value string) bool {
	if len(value) != 36 || value[8] != '-' || value[13] != '-' || value[18] != '-' || value[23] != '-' {
		return false
	}
	hexText := strings.ReplaceAll(value, "-", "")
	decoded, err := hex.DecodeString(hexText)
	return err == nil && len(decoded) == 16
}

func validDisplayText(value string, maxBytes int, allowEmpty bool) bool {
	if len(value) > maxBytes || (!allowEmpty && value == "") {
		return false
	}
	for _, character := range value {
		if unicode.IsControl(character) {
			return false
		}
	}
	return true
}

func descriptorInvalid(message string) error {
	return fmt.Errorf("invalid descriptor message: %s", message)
}

File: edge-node/core/publish/src/descriptor.rs
use serde::{Deserialize, Serialize};
use std::collections::HashSet;

use crate::PublishError;

pub const DESCRIPTOR_SCHEMA_VERSION: u32 = 2;
pub const MAX_DESCRIPTOR_BYTES: usize = 1024 * 1024;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct DescriptorSnapshot {
    pub schema_version: u32,
    pub edge_node_id: String,
    pub ledger_epoch: String,
    pub descriptor_revision: u64,
    pub complete: bool,
    pub devices: Vec<DescriptorDevice>,
    pub signals: Vec<DescriptorSignal>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct DescriptorDevice {
    pub system_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub identifier: Option<String>,
    pub state: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub model_id: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct DescriptorSignal {
    pub series_key: String,
    pub system_id: String,
    pub measurement_key: String,
    pub channel_index: Option<i32>,
    pub variant: String,
    pub unit: Option<String>,
    pub value_type: String,
}

impl DescriptorSnapshot {
    pub fn decode_bounded(payload: &[u8]) -> Result<Self, PublishError> {
        if payload.len() > MAX_DESCRIPTOR_BYTES {
            return Err(PublishError::Invalid(format!(
                "descriptor snapshot exceeds {MAX_DESCRIPTOR_BYTES} encoded bytes"
            )));
        }
        let snapshot: Self = serde_json::from_slice(payload).map_err(|error| {
            PublishError::Invalid(format!("descriptor decoding failed: {error}"))
        })?;
        snapshot.validate()?;
        Ok(snapshot)
    }

    pub fn encode_bounded(&self) -> Result<Vec<u8>, PublishError> {
        self.validate()?;
        let payload = serde_json::to_vec(self).map_err(|error| {
            PublishError::Invalid(format!("descriptor encoding failed: {error}"))
        })?;
        if payload.len() > MAX_DESCRIPTOR_BYTES {
            return Err(PublishError::Invalid(format!(
                "descriptor snapshot exceeds {MAX_DESCRIPTOR_BYTES} encoded bytes"
            )));
        }
        Ok(payload)
    }

    fn validate(&self) -> Result<(), PublishError> {
        if self.schema_version != DESCRIPTOR_SCHEMA_VERSION || !self.complete {
            return invalid("schema_version");
        }
        validate_descriptor_topic_segment(&self.edge_node_id)?;
        validate_descriptor_identity(&self.ledger_epoch)?;
        if self.descriptor_revision == 0 || self.descriptor_revision > i64::MAX as u64 {
            return invalid("descriptor_revision");
        }

        let mut device_ids = HashSet::with_capacity(self.devices.len());
        for device in &self.devices {
            validate_system_id(&device.system_id)?;
            if !device_ids.insert(device.system_id.as_str()) {
                return invalid("duplicate descriptor device system_id");
            }
            if !matches!(device.state.as_str(), "quarantined" | "active" | "retired") {
                return invalid("unsupported descriptor device state");
            }
            if let Some(identifier) = &device.identifier
                && (identifier.is_empty()
                    || identifier.len() > 64
                    || identifier.chars().any(char::is_control))
            {
                return invalid("invalid descriptor device identifier");
            }
            if let Some(model_id) = &device.model_id
                && !iotkit_core_ledger::is_valid_model_id(model_id)
            {
                return invalid("invalid descriptor device model_id");
            }
        }

        let mut series_keys = HashSet::with_capacity(self.signals.len());
        for signal in &self.signals {
            validate_system_id(&signal.system_id)?;
            if !device_ids.contains(signal.system_id.as_str()) {
                return invalid("descriptor signal references an unknown device");
            }
            if !valid_measurement_key(&signal.measurement_key) {
                return invalid("measurement_key");
            }
            if signal.channel_index.is_some_and(|channel| channel < 0)
                || signal.variant.is_empty()
                || signal.variant.contains(':')
            {
                return invalid("identity_boundary");
            }
            if signal.variant.chars().any(char::is_control) {
                return invalid("identity_control");
            }
            let channel = signal
                .channel_index
                .map(|value| value.to_string())
                .unwrap_or_else(|| "na".into());
            let expected = format!(
                "{}:{}:{}:{}",
                signal.system_id, signal.measurement_key, channel, signal.variant
            );
            if signal.series_key != expected {
                return invalid("descriptor series_key does not match signal identity");
            }
            if !series_keys.insert(signal.series_key.as_str()) {
                return invalid("duplicate descriptor signal series_key");
            }
            if !matches!(
                signal.value_type.as_str(),
                "float" | "int" | "bool" | "record"
            ) {
                return invalid("unsupported descriptor signal value_type");
            }
            if let Some(unit) = &signal.unit
                && (unit.len() > 128 || unit.chars().any(char::is_control))
            {
                return invalid("invalid descriptor signal unit");
            }
        }
        Ok(())
    }
}

fn validate_system_id(value: &str) -> Result<(), PublishError> {
    let parsed = uuid::Uuid::parse_str(value)
        .map_err(|_| PublishError::Invalid("noncanonical_uuid".into()))?;
    if parsed.hyphenated().to_string() != value {
        return invalid("noncanonical_uuid");
    }
    Ok(())
}

fn validate_descriptor_topic_segment(value: &str) -> Result<(), PublishError> {
    validate_descriptor_identity(value)?;
    if value.contains(['/', '+', '#']) {
        return invalid("identity_boundary");
    }
    Ok(())
}

fn validate_descriptor_identity(value: &str) -> Result<(), PublishError> {
    if value.is_empty() || value.len() > 255 || value.contains(':') {
        return invalid("identity_boundary");
    }
    if value.chars().any(char::is_control) {
        return invalid("identity_control");
    }
    Ok(())
}

fn valid_measurement_key(value: &str) -> bool {
    !value.is_empty()
        && value.len() <= 64
        && value.split('.').all(|segment| {
            let mut bytes = segment.bytes();
            bytes.next().is_some_and(|byte| byte.is_ascii_lowercase())
                && bytes
                    .all(|byte| byte.is_ascii_lowercase() || byte.is_ascii_digit() || byte == b'_')
        })
}

fn invalid<T>(message: &str) -> Result<T, PublishError> {
    Err(PublishError::Invalid(message.into()))
}
